src/libmeasurement_kit/libevent/dns_impl.hpp

```diff
@@ -53,37 +54,66 @@ class QueryContext : public NonMovable, public NonCopyable {
     }
 };
 
+struct evdns_base_deleter {
+    void operator()(evdns_base* p) {
+        constexpr int fail_requests = 1;
+        evdns_base_free(p, fail_requests);
+    }
+};
+
+struct evaddrinfo_deleter {
+    void operator()(evutil_addrinfo* p) {
+        evutil_freeaddrinfo(p);
+    }
+};
+
+using evdns_base_uptr = std::unique_ptr<evdns_base, evdns_base_deleter>;
+using evaddrinfo_uptr = std::unique_ptr<evutil_addrinfo, evaddrinfo_deleter>;
+
 template <MK_MOCK(evdns_base_new), MK_MOCK(evdns_base_nameserver_ip_add),
         MK_MOCK(evdns_base_free), MK_MOCK(evdns_base_set_option)>
 static inline evdns_base *create_evdns_base(
         Settings settings, Var<Reactor> reactor = Reactor::global()) {
 
-    evdns_base *base;
     event_base *evb = reactor->get_event_base();
+    const int initialize_nameservers = settings.count("dns/nameserver") ? 0 : 1;
+    evdns_base_uptr base(evdns_base_new(evb, initialize_nameservers));
+    if (!base) {
+        throw std::bad_alloc();
```

Comment:
Note to self: as part of future refactoring, I'd like this function to return `Var<event_base>` or possibly even better if it would return a `ErrorOr<Var<event_base>>` rather than throwing.